src/common/utils/time/time.ts

```diff
@@ -59,44 +49,44 @@ export function exclusiveToInclusiveEnd(exclusiveEnd: Date): Date {
-
-  let showingYear = true;
-  let formatted: string;
-  if (startWallTime.year() !== endWallTimeInclusive.year()) {
-    formatted = [startWallTime.format(FORMAT_WITH_YEAR), endWallTimeInclusive.format(FORMAT_WITH_YEAR)].join(" - ");
-  } else {
-    showingYear = displayYear === DisplayYear.ALWAYS ||
-      (displayYear === DisplayYear.IF_DIFF && !isCurrentYear(endWallTimeInclusive.year(), timezone));
-    const fmt = showingYear ? FORMAT_WITH_YEAR : FORMAT_WITHOUT_YEAR;
-    if (startWallTime.month() !== endWallTimeInclusive.month() || startWallTime.date() !== endWallTimeInclusive.date()) {
-      formatted = [startWallTime.format(FORMAT_WITHOUT_YEAR), endWallTimeInclusive.format(fmt)].join(" - ");
-    } else {
-      formatted = startWallTime.format(fmt);
-    }
+function formatDate(start: Moment, end: Moment, showingYear: boolean): string {
+  if (start.year() !== end.year()) {
+    return [start.format(FORMAT_WITH_YEAR), end.format(FORMAT_WITH_YEAR)].join(" - ");
   }
+  const fmt = showingYear ? FORMAT_WITH_YEAR : FORMAT_WITHOUT_YEAR;
+  if (start.month() !== end.month() || start.date() !== end.date()) {
+    return [start.format(FORMAT_WITHOUT_YEAR), end.format(fmt)].join(" - ");
+  }
+  return start.format(fmt);
+}
 
-  if (startWallTime.hour() || startWallTime.minute() || endWallTime.hour() || endWallTime.minute()) {
-    formatted += (showingYear ? " " : ", ");
-
-    let startTimeStr = formatTimeOfDay(startWallTime).toLowerCase();
-    const endTimeStr = formatTimeOfDay(endWallTime).toLowerCase();
+function formatTime(start: Moment, end: Moment): string {
+  if (!start.hour() && !end.hour() && !start.minute() && !end.minute()) return null;
+  return `${start.format(FORMAT_TIME_OF_DAY_WITH_MINUTES)} - ${end.format(FORMAT_TIME_OF_DAY_WITH_MINUTES)}`;
+}
 
-    if (startTimeStr === endTimeStr) {
-      formatted += startTimeStr;
-    } else {
-      if (startTimeStr.substr(-2) === endTimeStr.substr(-2)) {
-        startTimeStr = startTimeStr.substr(0, startTimeStr.length - 2);
-      }
-      formatted += [startTimeStr, endTimeStr].join("-");
-    }
+function showYear(displayYear: DisplayYear, startYear: number, endYear: number): boolean {
+  switch (displayYear) {
+    case DisplayYear.NEVER:
+      return false;
+    case DisplayYear.ALWAYS:
+      return true;
+    case DisplayYear.IF_DIFF:
+      return startYear !== endYear || endYear !== moment().year();
   }
+}
 
-  return formatted;
+export function formatTimeRange(range: TimeRange, timezone: Timezone, displayYear: DisplayYear): string {
+  const start = moment.tz(range.start, timezone.toString());
+  const end = moment.tz(range.end, timezone.toString());
+  const endInclusive = getEndWallTimeInclusive(range.end, timezone);
+
+  const showingYear = showYear(displayYear, start.year(), end.year());
+  const date = formatDate(start, endInclusive, showingYear);
+  const time = formatTime(start, end);
+
+  if (!time) return date;
```

Comment:
It should be checked against start/end but not against formatted time.

Reply:
Something like 

```
const shouldDisplayTime = f(start, end)
if (!shouldDisplayTime) return date;
return f(date, formatTime(start, end))
```

?